src/controller_examples/rabbitmq_controller/proof/helper_invariants/validation.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use crate::external_api::spec::{EmptyAPI, EmptyTypeView};
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, config_map::*, dynamic::*, error::*, owner_reference::*, resource::*,
+    stateful_set::*,
+};
+use crate::kubernetes_cluster::spec::{
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::rabbitmq_controller::{
+    common::*,
+    proof::{
+        helper_invariants::{owner_ref::*, predicate::*, proof::*},
+        predicate::*,
+        resource::*,
+    },
+    spec::{reconciler::*, resource::*, types::*},
+};
+use crate::reconciler::spec::{reconciler::*, resource_builder::*};
+use crate::temporal_logic::{defs::*, rules::*};
+use crate::vstd_ext::{multiset_lib, seq_lib, string_view::*};
+use vstd::{multiset::*, prelude::*, string::*};
+
+verus! {
+
+pub open spec fn stateful_set_satisfies_unchangeable(obj: DynamicObjectView, rabbitmq: RabbitmqClusterView) -> bool {
```

Comment:
Could you tell me why do we have lemmas specialized for stateful set here?

Reply:
I am not separating stateful set--you can see `lemma_always_object_in_etcd_satisfies_unchangeable` applies to all subresources. Maybe the naming here is a bit confusing.